### PR TITLE
Ports TG's BSQL library

### PR DESCRIPTION
**Pull request by ShiftyRail**
## What is it

Using TG's asynchronous BSQL library instead of our current SQL library, which is essentially just a wrapper for BYOND's native way to handle SQL, in order to stop server crashes.

BSQL's licensing is MIT and included in the library.

Thanks to Cyberboss and TGstation for this tool.

## Why it's good

SQL operations slow down the server quite a lot, can even cause crashes. This new library fixes that. I'm told it has its own crashes and little peculiarities, but probably (hopefully) not on the level of the current one.

Profiling BSQL did not show anything abnormal locally, unlike `DBQuery/proc/Execute` which gets on top of the profiler very often.

## Technical details

BSQL is a dll which handles the SQL calls and returns the values as a list of rows.
An API let us communicate and debug BSQL (somewhat) easily.
This API is included in a subsystem (initialized before everything else), which counts the active queries, alerts admins if they are slow or timing out, and provides an extra wrapper or ease of coding.

Unlike the current SQL library, this one is non-blocking and should not crash the server as often as the old one does. (there is absolutely no way to stop or defer a database call with our current library, which explains why its calls are ALWAYS overtime).

## What I did

- Un-TGify the code, making it compatible with /vg/.
- Change queries to use the new `dbcore` subsystem.
- Made sure an helpful error message is thrown when a query fails, and the relevant proc is stopped in order to let the subsystem do its work.
- Made sure **all** queries are properly `qdel`d after they are done. This is important, because without that, the queries are left hanging and can mess with the subsystem.
- Add a few migrations to make sure you can test it locally.
- Made the library computer use the feedback database and made a migration for that.
- Removed an exploit which would allow an admin with +VAREDIT or +DEBUG to do [redacted].

## 99% tested

I used WAMP to set up a local DB and see how well the library interacted with it. I think it does ok.

### Things that I have tested :
[x] Logging connections
[x] Admin sessions
[x] Player connections
[x] Bans
[x] Jobbans
[x] Server polls
[x] Database-side admin logs
[x] Adding and removing admins, adding and removing admin flags
[x] SQL migration controller (added a few migrations which were missing).
[x] Leaving the server on for a while and seeing if queries were undeleted
[x] Library

### Things that I couldn't test

[] Banned players are correctly detected by byond as banned, but they can still connect to the server. My guess is that it has to do with localhost as an IP.

### To note

The library as it is coded right now isn't compatible with BSQL. Either Pomf must migrate the existing table to the `feedback`database, or the library will be purged. (again).

### CODE REVIEW

**Comment by ghost** [comment]
Wouldn't it be better to use something like https://github.com/tgstation/rust-g/pull/25?

BSQL is buggy and prone to crashing. Also only like two people understand BSQL's code.

**Comment by ShiftyRail** [comment]
I see. I looked at BSQL because it had been already worked on by another /vg/ maintainer and I got more or less the gist of the DM part.

I am by no means qualified enough to judge how the C++ or RUST involved is working. I'm mostly here trying to figure out the DM and DM integration part.

If that `rust-g` is working and stable enough, then I'll happily change the DLLs.

**Comment by ghost** [comment]
The problem with my rust-g SQL implementation is that it uses parameterized queries - there is no sanitation function. Although parameterized queries is more or less the "right" way to do SQL anyways, you'd still need to refactor every single SQL query nonetheless.

The upside is that it's safe Rust code - so less/no crashing. Also easier to maintain.

**Comment by ShiftyRail** [comment]
>you'd still need to refactor every single SQL query nonetheless.

That looks painful. Every single one of our query is more or less a raw string with DM coding filling in the blanks.
I think I see what you want me to do, but I'm not sure if I can do it correctly. Let's first try what "worked" for TG and then look at the alternative.

**Comment by ShiftyRail** [comment]
This is now officially 25% tested.
It works, it connects to databases, and it even logs connections as needed

image

However, it always strips the first and last character in all fields, so it throws errors when banning, promoting admins, among a few other things.
I think this issue prevents the SQL migration controller from properly working too.
But we're almost there.

File: code/defines/procs/statistics.dm [hunk 1]
```diff
@@ -99,11 +99,11 @@ proc/sql_report_cyborg_death(var/mob/living/silicon/robot/H)
 	var/sqltime = time2text(world.realtime, "YYYY-MM-DD hh:mm:ss")
 	var/coord = "[H.x], [H.y], [H.z]"
 //	to_chat(world, "INSERT INTO death (name, byondkey, job, special, pod, tod, laname, lakey, gender, bruteloss, fireloss, brainloss, oxyloss) VALUES ('[sqlname]', '[sqlkey]', '[sqljob]', '[sqlspecial]', '[sqlpod]', '[sqltime]', '[laname]', '[lakey]', '[H.gender]', [H.bruteloss], [H.getFireLoss()], [H.brainloss], [H.getOxyLoss()])")
-	establish_db_connection()
-	if(!dbcon.IsConnected())
+	SSdbcore.Connect()
+	if(!SSdbcore.IsConnected())
```
Comment (Cyberboss) [inline code/defines/procs/statistics.dm]: succinct version
```suggestion
	if(!SSdbcore.Connect())
```

**Comment by ShiftyRail** [comment]
image

it just works

**Comment by ShiftyRail** [comment]
Unfortunatly, even though `qdel`ing stuff is what TG does, I still get runtimes for `"Operation identifier does not exist!"`. even though the SQL is correctly identified and running well.

I have some DM-only problems with polls and the library/custom items (mostly because the old database somehow isn't called by the right name).

SQL migration still isn't working. Maybe because it was meant for MySQL and I'm using MariaDB.

**Comment by ghost** [comment]
pretty sure everyone uses MariaDB

**Comment by ShiftyRail** [comment]
Back when I learned PHP in 2007 they didn't..........

**Comment by ShiftyRail** [comment]
image

Jobbans work too. I can't seem to ban myself on localhost (probably a byond thing?), but the SQL is executed and it correctly detects that the guy is banned.

I didn't test it ULTRA ROBUSTLY, but everything seems to work so far (polls, library, bans, admins, permissions, admin logs, connection logs, etc.).

I'm going to give it another round of testing this week-end, and then I /think/ it could reasonably be live-tested on the test server for a few rounds.

**Comment by d3athrow** [comment]
>[] Banned players are correctly detected by byond as banned, but they can still connect to the server. My guess is that it has to do with localhost as an IP.

@ShiftyRail If you forward the port on your router you can connect to your local server using your remote IP.

File: code/modules/migrations/SS13/011-port-library-to-feedback-db.dm [hunk 2]
```diff
@@ -0,0 +1,30 @@
+/datum/migration/mysql/ss13/_011
+	id = 11
+	name = "Library in feedback DB"
+
+/datum/migration/mysql/ss13/_011/up()
+	if(!hasTable("library"))
+		return execute({"CREATE  TABLE IF NOT EXISTS `tgstation`.`library` (
```
Comment (d3athrow) [inline code/modules/migrations/SS13/011-port-library-to-feedback-db.dm]: Extra space between CREATE and TABLE. Also our db is called feedback, not tgstation

Comment (ShiftyRail) [reply code/modules/migrations/SS13/011-port-library-to-feedback-db.dm]: Got it. I think I can even remove the db name, as we're going to only get one anyways.

File: __DEFINES/subsystem.dm [hunk 3]
```diff
@@ -2,6 +2,8 @@
 // All in one file so it's easier to see what everything is relative to.
 
 #define SS_INIT_TICKER_SPAWN       999
+#define SS_INIT_DBCORE			   900
```
Comment (ShiftyRail) [inline __DEFINES/subsystem.dm:5]: Deprecated, should be removed

File: code/__HELPERS/logging.dm [hunk 4]
```diff
@@ -54,8 +54,19 @@
 		if(C.prefs.toggles & CHAT_DEBUGLOGS)
 			to_chat(C, "DEBUG: [text]")
 
-
-
+/proc/log_sql(text)
+	world.log << "\[[time_stamp()]]SQL: [text]"
```
Comment (ShiftyRail) [inline code/__HELPERS/logging.dm]: duplicate logs. I used those for easier debug via DD, but they're not needed.

**Comment by ShiftyRail** [comment]
Secret repo PR done

**Comment by ShiftyRail** [comment]
Thinks to look for still :

- server polls don't show up. This might be DM-side as I don't see any runtime or SQL error for it.
- one roundstart runtime related to migration. Doesn't affect much, but should be fixed.
- `SessID` throws an error when it shouldn't.
- `SQL: Undeleted query: "" LA: Created LAT: 325` for some reason.

**Pull request merged**